contract/src/msgs/sorted_set.rs

```diff
@@ -0,0 +1,90 @@
+use seda_common::msgs::data_requests::DataRequest;
+
+use super::*;
+
+/// IndexKey is a tuple of (gas_price, height, dr_id)
+pub type IndexKey = (u128, u64, Hash);
+
+/// A structure to store a sorted set of data requests by the `IndexKey`
+pub struct SortedSet<'a> {
+    pub len:            Item<u32>,
+    /// Used to store information about the data request by the `IndexKey` so it can be sorted
+    pub index:          Map<IndexKey, ()>,
+    /// Used to store the `IndexKey` by the `Hash` of the data request
+    pub dr_id_to_index: Map<&'a Hash, IndexKey>,
+}
+
+impl SortedSet<'_> {
+    pub fn initialize(&self, store: &mut dyn Storage) -> StdResult<()> {
+        self.len.save(store, &0)?;
+        Ok(())
+    }
+
+    pub fn len(&self, store: &dyn Storage) -> StdResult<u32> {
+        self.len.load(store)
+    }
+
+    pub fn has(&self, store: &dyn Storage, dr_id: &Hash) -> bool {
+        self.dr_id_to_index.has(store, dr_id)
+    }
+
+    pub fn has_index(&self, store: &dyn Storage, index: IndexKey) -> bool {
+        self.index.has(store, index)
+    }
+
+    pub fn add(&self, store: &mut dyn Storage, dr_id: &Hash, dr: DataRequest) -> StdResult<()> {
+        if self.has(store, dr_id) {
+            return Err(StdError::generic_err("Key already exists"));
+        }
+
+        let gas_price: u128 = dr.gas_price.into();
+        let height: u64 = u64::MAX - dr.height;
```

Comment:
I think maybe it makes sense to also add Mario's comment here on why we 'convert' the height?

Reply:
I did. It's just in a different spot: https://github.com/sedaprotocol/seda-chain-contracts/pull/279/commits/f816c8b97cb05fecbb002e4e0bc9c37c97e42fea#diff-b0a0846865b5f60cb05dc4bf43c5475b4f66bcc69f89713307424a0fadada425R252-R254.

Should the comment be in both places? Or ideally we should refactor so it's done in one place

Reply:
If it's not too much hassle I think having a helper function or something would be great!